### PR TITLE
Refactor denoising

### PR DESCRIPTION
**Pull request by KaiWaldrant**

### CODE REVIEW

**Review by rcannood — CHANGES_REQUESTED** [review]
Making some good progress here!

I think the most important comment is to get the unit tests working, as it will be able to tell you when certain things aren't working.

File: src/denoising/data_processing/split_data/script.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+#   copy adata to train_set, test_set
+
+output_train = adata[:, ~is_missing].copy()
+del output_train.layers["counts"]
+output_train.layers["counts"] = scipy.sparse.csr_matrix(X_train).astype(float)
+
+output_test = adata[:, ~is_missing].copy()
+del output_test.layers["counts"]
+output_test.layers["counts"] = scipy.sparse.csr_matrix(X_test).astype(float)
```
[the remark this answers is not in the document]
Comment (KaiWaldrant) [reply src/denoising/data_processing/split_data/script.py:64]: I Allways get the error that the matrix size is not the same as it expected in the counts layer. that's why I delete the layer first.

File: src/denoising/control_methods/no_denoising/config.vsh.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  namespace: "denoising/control_methods"
+  description: "negative control by copying train counts"
+  info:
+    type: negative_control
+    label: no denoising
+    # paper_name: "Molecular Cross-Validation for Single-Cell RNA-seq""
+    # paper_url: "https://doi.org/10.1101/786269"
+    # paper_year: 2019
+    paper_doi: "10.1101/786269"
+    code_url: "https://github.com/czbiohub/molecular-cross-validation"
```
[the remark this answers is not in the document]
Comment (KaiWaldrant) [reply src/denoising/control_methods/no_denoising/config.vsh.yaml:13]: was a copy paste from the v1 script. but you are correct as it is not even related.

File: src/denoising/methods/dca/script.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+## VIASH END
+
+print("load input data")
+input_train = ad.read_h5ad(par['input_train'])
+
+print("process data")
+
+# make adata object with train counts
+# run DCA
+dca(input_train, epochs=par["epochs"])
```
[the remark this answers is not in the document]
Comment (KaiWaldrant) [reply src/denoising/methods/dca/script.py:24]: DCA still needs to be worked on as it has package issues and was not updated with the correct layers etc

File: src/denoising/methods/knn_smoothing/script.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    'functionality_name': 'foo',
+}
+## VIASH END
+
+print("Load input data")
+input_train = ad.read_h5ad(par["input_train"])
+
+print("process data")
+
+X = input_train.X.transpose().toarray()
```
[the remark this answers is not in the document]
Comment (KaiWaldrant) [reply src/denoising/methods/knn_smoothing/script.py:21]: Same here. script needs to be updated to the current way of working. Also has python package issues.

File: src/denoising/data_processing/split_data/script.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rcannood) [reply src/denoising/data_processing/split_data/script.py:64]: I see, fair enough!

Maybe it makes sense to make the output objects from scratch anyway, so ensure that no extra fields get passed to the methods anyways

**Comment by rcannood** [comment]
@KaiWaldrant Could you list in the Changelog what this PR adds to the repository? When applicable, could you also note what the differences are?

File: src/denoising/methods/dca/script.py [hunk 5]
```diff
@@ -18,7 +18,6 @@
 
 print("move layer to X")
 input_train.X = input_train.layers["counts"]
-del input_train.layers["counts"]
```
Comment (rcannood) [inline src/denoising/methods/dca/script.py]: Oops, my bad

**Review by rcannood — APPROVED** [review]
LGTM, as long as you can add a changelog afterwards @KaiWaldrant

**Pull request merged**